File: Battelle.EPA.WideAreaDecon.Client/src/implementations/jobs/JobRequest.ts
import JobStatus from '@/enums/jobs/jobStatus';
import IJobRequest from '@/interfaces/jobs/IJobRequest';
import { JsonProperty, Serializable } from 'typescript-json-serializer';
import ParameterWrapperList from '@/implementations/parameter/ParameterWrapperList';
import IJobResultRealization from '@/interfaces/jobs/results/IJobResultRealization';
import { ScenarioDefinitionMode } from '@/types';

@Serializable()
export default class JobRequest implements IJobRequest {
  @JsonProperty({
    // don't include in serialization
    onSerialize: () => undefined,
  })
  id: string;

  @JsonProperty()
  status: JobStatus;

  @JsonProperty()
  progress: number;

  @JsonProperty()
  defineScenario: ParameterWrapperList;

  @JsonProperty()
  modifyParameter: ParameterWrapperList;

  @JsonProperty()
  numberRealizations: number;

  @JsonProperty()
  seed1: number;

  @JsonProperty()
  seed2: number;

  @JsonProperty()
  results: IJobResultRealization[];

  @JsonProperty()
  definitionMode: ScenarioDefinitionMode | null;

  constructor(
    status: JobStatus = JobStatus.new,
    defineScenario: ParameterWrapperList,
    modifyParameter: ParameterWrapperList,
    numberRealizations: number,
    seed1: number,
    seed2: number,
    definitionMode: ScenarioDefinitionMode | null = null,
  ) {
    this.id = '';
    this.status = status;
    this.progress = 0;
    this.defineScenario = defineScenario;
    this.modifyParameter = modifyParameter;
    this.numberRealizations = numberRealizations;
    this.seed1 = seed1;
    this.seed2 = seed2;
    this.results = [];
    this.definitionMode = definitionMode;
  }
}

File: Battelle.EPA.WideAreaDecon.API/Models/Job/JobRequest.cs
﻿using System;
using System.Collections.Generic;
using Battelle.EPA.WideAreaDecon.API.Enumeration.Job;
using Battelle.EPA.WideAreaDecon.InterfaceData.Models.Parameter;
using Newtonsoft.Json;
using Battelle.EPA.WideAreaDecon.InterfaceData.Models.Results;
using Newtonsoft.Json.Converters;
using Battelle.EPA.WideAreaDecon.InterfaceData.Enumeration.Job;

namespace Battelle.EPA.WideAreaDecon.API.Models.Job
{
    /// <summary>
    /// Contains all the information related to a job
    /// </summary>
    public class JobRequest
    {
        [JsonProperty]
        public Guid Id { get; set; }

        public double Progress { get; set; }

        public JobStatus Status { get; set; } = JobStatus.New;

        [JsonProperty(Required = Required.Always)]
        public ParameterList DefineScenario { get; set; }

        [JsonProperty(Required = Required.Always)]
        public ParameterList ModifyParameter { get; set; }

        [JsonProperty(Required = Required.Always)]
        public int NumberRealizations { get; set; }

        public List<JobResults> Results { get; set; }

        [JsonProperty]
        public ulong Seed1 { get; set; }

        [JsonProperty]
        public ulong Seed2 { get; set; }

        [JsonProperty]
        [JsonConverter(typeof(StringEnumConverter))]
        public ScenarioDefinitionMode DefinitionMode { get; set; }


        public bool IsGeospatialDefinition => DefinitionMode == ScenarioDefinitionMode.Geospatial;
    }
}